File: functions/src/shared/canonicalEvent.ts
/**
 * Canonical event schema — the same hard contract as the Flutter client's
 * `lib/models/calendar_event.dart` (field names match exactly, so a
 * Firestore document looks identical regardless of whether it was written
 * by the app or by this Cloud Function). Timestamps are ISO-8601 UTC
 * strings; Firestore driver conversion to/from `Timestamp` happens at the
 * repository boundary, not here, so this stays a plain, easily-testable
 * data shape.
 */
export type EventSource =
  | "manual"
  | "upload"
  | "screenshot"
  | "google"
  | "outlook"
  | "icloud"
  | "sports";

export type EventStatus = "active" | "pendingConflict";

export type EventImportance = "locked" | "flexible";

export type ConflictRole = "original" | "new";

export interface CanonicalEventData {
  title: string;
  location: string | null;
  /** Always UTC, ISO-8601. Normalize at ingestion; never store local time. */
  start: string;
  end: string;
  source: EventSource;
  sourceId: string | null;
  /** Followed team ids represented by a sports event. */
  sportsTeamIds: string[] | null;
  status: EventStatus;
  /** Only direct user action or Step 5's routing logic may ever set this. */
  tag: string | null;
  importance: EventImportance;
  notes: string | null;
  attachments: string[] | null;
  repeat: Record<string, unknown> | null;
  reminders: string[] | null;
  conflictGroupId: string | null;
  conflictRole: ConflictRole | null;
}

export interface CanonicalEvent extends CanonicalEventData {
  id: string;
}

/** Defaults for fields a source adapter (this one included) never sets. */
export function newSourceEvent(
  fields: Pick<
    CanonicalEventData,
    "title" | "location" | "start" | "end" | "source" | "sourceId" | "notes"
  >,
): CanonicalEventData {
  return {
    ...fields,
    sportsTeamIds: null,
    status: "active",
    tag: null,
    importance: "flexible",
    attachments: null,
    repeat: null,
    reminders: null,
    conflictGroupId: null,
    conflictRole: null,
  };
}

File: scripts/sports_poller/canonical_event.py
"""Canonical event schema, in Python.

A field-for-field port of the Dart client's `lib/models/calendar_event.dart`
and its TypeScript counterpart `functions/src/shared/canonicalEvent.ts` --
kept in sync with both so a Firestore document this poller writes is
indistinguishable from one either of those write. Timestamps are ISO-8601
UTC strings; conversion to/from Firestore's native timestamp type happens
at the repository boundary (`event_repository.py`), not here, so this stays
a plain, easily-testable data shape.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class CanonicalEventData:
    title: str
    location: Optional[str]
    # Always UTC, ISO-8601. Normalize at ingestion; never store local time.
    start: str
    end: str
    source: str
    source_id: Optional[str]
    sports_team_ids: Optional[list[str]]
    status: str
    # Only direct user action or Step 5's routing logic may ever set this.
    tag: Optional[str]
    importance: str
    notes: Optional[str]
    attachments: Optional[list]
    repeat: Optional[dict]
    reminders: Optional[list]
    conflict_group_id: Optional[str]
    conflict_role: Optional[str]


@dataclass(frozen=True)
class CanonicalEvent(CanonicalEventData):
    id: str


def new_source_event(
    *,
    title: str,
    location: Optional[str],
    start: str,
    end: str,
    source: str,
    source_id: Optional[str],
    notes: Optional[str],
    sports_team_ids: Optional[list[str]] = None,
) -> CanonicalEventData:
    """Defaults for fields a source adapter (this poller included) never sets."""
    return CanonicalEventData(
        title=title,
        location=location,
        start=start,
        end=end,
        source=source,
        source_id=source_id,
        sports_team_ids=sports_team_ids,
        status="active",
        tag=None,
        importance="flexible",
        notes=notes,
        attachments=None,
        repeat=None,
        reminders=None,
        conflict_group_id=None,
        conflict_role=None,
    )
